src/charm.py

```diff
@@ -633,7 +669,66 @@ def _get_backup_key(self) -> str:
 
-    def _get_ca_cert_location_in_charm(self) -> str:
+    def _on_new_access_certificate_joined(self, event: RelationJoinedEvent):
+        if not self._ingress_address:
+            raise RuntimeError("Tried to process relation join before peer relation.")
+        if not (private_key := self._get_private_key_from_workload().encode()):
+            private_key = generate_private_key()
+        new_csr = generate_csr(private_key=private_key, subject=self._ingress_address)
+        self._push_csr_to_workload(new_csr.decode())
+        self._push_unit_pk_to_workload(private_key=private_key.decode())
+        self.tls_certificate_access.request_certificate_creation(
+            certificate_signing_request=new_csr
+        )
+        logger.info(
+            "New certificate requested from %s for unit %s",
+            event.relation.name,
+            self.unit.name,
+        )
+
+    def _on_new_access_certificate_available(self, event: CertificateAvailableEvent):
+        private_key = self._get_private_key_from_workload()
+        csr = self._get_csr_from_workload()
+        if not private_key or csr != event.certificate_signing_request:
+            return
+        self._push_unit_certificate_to_workload(certificate=event.certificate)
+        self._push_ca_certificate_to_workload(certificate=event.ca)
+        self._container.send_signal(SIGHUP, self._container_name)
+
+    def _on_access_certificate_expiring(self, event: CertificateExpiringEvent):
+        private_key = self._get_private_key_from_workload()
+        old_csr = self._get_csr_from_workload()
+        old_cert = self._get_cert_from_workload()
+        if event.certificate != old_cert:
+            return
+        if not self._ingress_address:
+            raise RuntimeError("Peer relation somehow unavailable. Something is very wrong")
+        new_csr = generate_csr(private_key=private_key.encode(), subject=self._ingress_address)
+        self._push_csr_to_workload(new_csr.decode())
+        self.tls_certificate_access.request_certificate_renewal(
+            old_certificate_signing_request=old_csr.encode(),
+            new_certificate_signing_request=new_csr,
+        )
+        logger.info("Sent a certificate renewal request for %s.", self.unit.name)
+
+    def _on_access_certificate_invalidated(
+        self, event: Union[AllCertificatesInvalidatedEvent, CertificateInvalidatedEvent]
+    ):
+        if (
+            hasattr(event, "certificate_signing_request")
+            and event.certificate_signing_request != self._get_csr_from_workload()
+        ):
+            return
+        self._remove_all_certs_from_workload()
+        self._restore_self_signed_certs_to_workload()
+        self._container.send_signal(SIGHUP, self._container_name)
+
+    def _get_authentication_details_for_vault(self) -> Dict[str, str]:
+        ca_cert = self._get_ca_cert_path_in_charm()
+        cert, pk = self._get_cert_and_pk_path_in_charm()
```

Comment:
Why the need for those 2 different methods, the location is the same for all those files.

Reply:
Since there's some scenarios where we replace the cert and pk and not the ca, and now with the tls access interface there's a chance that these files may not be there, I thought it was a good idea to split these up and error check them separately. I also attempted in this function to introduce a kind of facade in case in the future we need to support multiple different ways or ca,cert,pk locations for us and users to authenticate with vault, for example if we in the future need to differentiate different certificate folders for PKI.

I think there could be an issue with the names getting "paths" instead of the actual files, so people may assume that it's accessing a static variable instead of actually checking if the paths are valid also. If you agree with this I can change the names of the functions to get the files instead of the paths.